app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconPersister.kt

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2020 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.favicon
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.file.FileDeleter
+import com.duckduckgo.app.global.sha256
+import kotlinx.coroutines.NonCancellable
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileOutputStream
+
+interface FaviconPersister {
+    fun faviconFile(directory: String, subFolder: String, domain: String): File?
+    suspend fun store(directory: String, subFolder: String, bitmap: Bitmap, domain: String): File?
+    suspend fun copyToDirectory(file: File, directory: String, newSubfolder: String, newFilename: String)
+    suspend fun deleteAll(directory: String)
+    suspend fun deletePersistedFavicon(domain: String)
+    suspend fun deleteFaviconsForSubfolder(directory: String, subFolder: String, domain: String?)
+}
+
+class FileBasedFaviconPersister(
+    val context: Context,
+    private val fileDeleter: FileDeleter,
+    private val dispatcherProvider: DispatcherProvider
+) : FaviconPersister {
+
+    override suspend fun deleteAll(directory: String) {
+        fileDeleter.deleteDirectory(faviconDirectory(directory))
+    }
+
+    override fun faviconFile(directory: String, subFolder: String, domain: String): File? {
+        val file = fileForFavicon(directory, subFolder, domain)
+        return if (file.exists()) {
+            file
+        } else {
+            null
+        }
+    }
+
+    override suspend fun copyToDirectory(file: File, directory: String, newSubfolder: String, newFilename: String) {
+        withContext(dispatcherProvider.io()) {
+            val persistedFile = fileForFavicon(directory, newSubfolder, newFilename)
+            file.copyTo(persistedFile, overwrite = true)
```

Comment:
given that this is writing files, should we `synchronize` this function too?

Reply:
Done!